### ISSUE TITLE
Is nn.Conv2d equivalent with Unfold + Matrix Multiplication + Fold ?

### ISSUE THREAD

**Issue by evinpinar** [pytorch/pytorch#47990]
## Is nn.Conv2d equivalent with Unfold + Matrix Multiplication + Fold

I have implemented a custom convolutional operator with Unfold+matrix multiplication+Fold as described in the pytorch documentation, under Unfold examples [here](https://pytorch.org/docs/stable/generated/torch.nn.Unfold.html):

`>>> # Convolution is equivalent with Unfold + Matrix Multiplication + Fold (or view to output shape)
>>> inp = torch.randn(1, 3, 10, 12)
>>> w = torch.randn(2, 3, 4, 5)
>>> inp_unf = torch.nn.functional.unfold(inp, (4, 5))
>>> out_unf = inp_unf.transpose(1, 2).matmul(w.view(w.size(0), -1).t()).transpose(1, 2)
>>> out = torch.nn.functional.fold(out_unf, (7, 8), (1, 1))
>>> # or equivalently (and avoiding a copy),
>>> # out = out_unf.view(1, 2, 7, 8)
>>> (torch.nn.functional.conv2d(inp, w) - out).abs().max()`

Ideally, this should be the same thing with the nn.Conv2d, however my simple experiments on a vision understanding task does not say so. What can be the reason? Are there any interesting optimizations happening on cuda implementation?

## To Reproduce

Here is my custom convolution:

`class CustomConv(nn.Module):

    def __init__(self, in_channel, out_channel, kernel_size=3, stride=1, padding=0, bias=True):
        super(CustomConv, self).__init__()
        self.k = kernel_size
        self.in_c = in_channel
        self.out_c = out_channel
        self.stride = stride
        self.padding = padding
        self.conv = nn.Parameter(torch.ones(out_channel, in_channel, kernel_size, kernel_size))
        nn.init.xavier_uniform_(self.conv, gain=nn.init.calculate_gain('relu'))
        #nn.init.kaiming_normal_(self.conv, mode='fan_out', nonlinearity='relu')

    def forward(self, inp):
        k = self.k
        stride = self.stride
        h_in, w_in = inp.shape[2], inp.shape[3]

        padding = self.padding  # + k//2
        batch_size = inp.shape[0]

        h_out = (h_in + 2 * padding - (k - 1) - 1) / stride + 1
        w_out = (w_in + 2 * padding - (k - 1) - 1) / stride + 1
        h_out, w_out = int(h_out), int(w_out)

        inp_unf = torch.nn.functional.unfold(inp, (k, k), padding=padding)
        out_unf = inp_unf.transpose(1, 2).matmul(self.conv.view(self.conv.size(0), -1).t()).transpose(1, 2)
        #out_ = out_unf.view(batch_size, self.out_c, h_out, w_out) # this is equivalent with fold.
        out_ = torch.nn.functional.fold(out_unf, (h_out, w_out), (1, 1))
`

## Expected behavior

I expected to be able to train my network with this CustomConv, instead of nn.Conv2d. But I cannot replicate the results.

## Environment

PyTorch version: 1.5.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.17.3

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 450.51.06
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.0
[pip3] torch==1.5.1
[pip3] torchfile==0.1.0
[pip3] torchnet==0.0.5.1
[pip3] torchvision==0.6.1

**Comment by zou3519** [comment]
The values should be the same. The first code snippet works on my machine:
```
    inp = torch.randn(1, 3, 10, 12)
    w = torch.randn(2, 3, 4, 5)
    inp_unf = torch.nn.functional.unfold(inp, (4, 5))
    out_unf = inp_unf.transpose(1, 2).matmul(w.view(w.size(0), -1).t()).transpose(1, 2)
    out = torch.nn.functional.fold(out_unf, (7, 8), (1, 1))
    # or equivalently (and avoiding a copy),
    out = out_unf.view(1, 2, 7, 8)
    (torch.nn.functional.conv2d(inp, w) - out).abs().max()    
```
Do you have an example where your convolution returns a result that is different from that of nn.Conv2d?

**Comment by evinpinar** [comment]
Thanks for the answer. They return the same thing, but I am not sure if they are optimized in a same way ie. gradient update on gpus. There is a misalignment there.

**Comment by DegardinBruno** [comment]
@evinpinar do you already got your answer? 
I'm running into the same issue, the output of a custom convolution (with fold and unfold) are the same as the Conv2d, but when training it seems that the weights are not optimized in the same way as the Conv2d.
Even with the default weight initialization of Conv2d.

**Comment by evinpinar** [comment]
In the end, I have added bias and uniform weight initialization: 
`if bias:
            self.bias = nn.Parameter(torch.zeros(out_channel).float(), requires_grad=True)

        stdv = 1. / math.sqrt(self.conv.size(1))
        self.conv.data.uniform_(-stdv, stdv)
        if bias:
            self.bias.data.uniform_(-stdv, stdv)`

So far, it is not working in exactly the same way, but it is good enough for custom convolution.

**Comment by DegardinBruno** [comment]
Thank you for your answer, I also managed to make it work.
However, I think I know one of the reasons that you are telling me that "it is not working in exactly the same way"...
I would recommend you initialize the weights similarly to the Pytorch default, as follows:



        # Weights and Bias
        self.weight = nn.Parameter(torch.Tensor(self.out_channels, self.out_channels, *self.kernel_size))
        self.bias   = nn.Parameter(torch.zeros(self.out_channels)) if bias else 0
        
        # Weights and Bias initialization
        nn.init.kaiming_uniform_(self.weight, a=math.sqrt(5))
        if self.bias is not 0:
            fan_in, _ = nn.init._calculate_fan_in_and_fan_out(self.weight)
            bound = 1 / math.sqrt(fan_in)
            nn.init.uniform_(self.bias, -bound, bound)

**Comment by ShoufaChen** [comment]
Hi, @DegardinBruno @evinpinar,

Any update? Do these two implementations behave in the same way now? 

Thanks.

**Comment by ShoufaChen** [comment]
> The values should be the same. The first code snippet works on my machine:
> 
> ```
>     inp = torch.randn(1, 3, 10, 12)
>     w = torch.randn(2, 3, 4, 5)
>     inp_unf = torch.nn.functional.unfold(inp, (4, 5))
>     out_unf = inp_unf.transpose(1, 2).matmul(w.view(w.size(0), -1).t()).transpose(1, 2)
>     out = torch.nn.functional.fold(out_unf, (7, 8), (1, 1))
>     # or equivalently (and avoiding a copy),
>     out = out_unf.view(1, 2, 7, 8)
>     (torch.nn.functional.conv2d(inp, w) - out).abs().max()    
> ```
> 
> Do you have an example where your convolution returns a result that is different from that of nn.Conv2d?

Hi, @zou3519 

In my case, I use `torch.allclose` to test these two outputs. The results show that these two outputs are **sometimes** same, **sometimes** not same.

Any hint about this phenomenon?

**Comment by zou3519** [comment]
@ShoufaChen what happens if you try to use torch.double? (e.g. `inp = inp.double(); w = w.double()`). This might be a numerical precision issue.

**Comment by ShoufaChen** [comment]
@zou3519 Thanks. Using `.double()` would always get `True` for `allclose`.

**Comment by pbelcak** [comment]
Hello everyone; I'm afraid this continues to be an issue, and I think there is a bug in the backward pass through `Unfold`. Crossposting from the [PyTorch Forums](https://discuss.pytorch.org/t/unfold-linear-fold-conv2d/201992):

> This has appeared both on the forums and in PyTorch issues before ([this one is still open](https://github.com/pytorch/pytorch/issues/47990)). Nevertheless, I still find that when training a "simulated" Conv2d using `Unfold, Transpose, Linear, Transpose, Fold`, the gradients are different to using just the "equivalent" Conv2d.
> 
> For a reproducible example, I provide a "simulated" VGGxx implementation and loss curves from corresponding runs (all parameters being equal).
> 
> To get the simulated VGG, copy the contents of the torchvision implementation and modify `make_features` as follows:
> 
> 
> ```
> def make_layers(cfg: List[Union[str, int]], batch_norm: bool = False) -> nn.Sequential:
>     layers: List[nn.Module] = []
>     in_channels = 3
>     pooling_depth: int = 0
>     for v in cfg:
>         if v == "M":
>             layers += [nn.MaxPool2d(kernel_size=2, stride=2)]
>             pooling_depth += 1
>         else:
>             v = cast(int, v)
>             w, h = 224 // (2**pooling_depth), 224 // (2**pooling_depth)
>             layers += [ nn.Unfold(kernel_size=3, padding=1, stride=1) ]
>             layers += [ Transpose(-1, -2) ]
>             layers += [ nn.Linear(in_channels * 3*3, v, bias=True) ]
>             layers += [ nn.ReLU(inplace=True) ]
>             layers += [ Transpose(-1, -2) ]
>             # layers += [ Reshape(-1, v, w, h) ]
>             layers += [ nn.Fold(output_size=(w, h), kernel_size=1, stride=1) ]
>             
>             in_channels = v
>     return nn.Sequential(*layers)
> ```
> 
> The training harness used is the one of [`torchvision/classification`](https://github.com/pytorch/vision/tree/main/references/classification#alexnet-and-vgg), and I follow the recipe given in the README.
> 
> Here are the losses for the first few epochs of VGG11 reference (blue) and simulated (yellow):
> 
> [IMAGE]
> 
> Note that I also tried training both in double precision and it didn't make a difference.
> 
> Am I missing something, or is there genuinely a bug in `Unfold/Fold` backward pass?
